core/storage/buffer_pool.rs

```diff
@@ -1,45 +1,534 @@
+        let inner = pool.inner_mut();
+        // Just store the IO handle, don't create arena yet
+        if inner.io.is_none() {
+            inner.io = Some(Arc::clone(io));
+        }
+        pool.clone()
+    }
+
+    /// Call when `[Database::db_state]` is initialized, providing the `page_size` to allocate
+    /// an arena for the pool. Before this call, the pool will use temporary buffers which are
+    /// cached in thread local storage.
+    pub fn finalize_with_page_size(&self, page_size: usize) -> crate::Result<Arc<Self>> {
+        let pool = BUFFER_POOL.get().expect("BufferPool must be initialized");
+        let inner = pool.inner_mut();
+        tracing::trace!("finalize page size called with size {page_size}");
+        if page_size != BufferPool::DEFAULT_PAGE_SIZE {
+            // so far we have handed out some temporary buffers, since the page size is not
+            // default, we need to clear the cache so they aren't reused for other operations.
+            TEMP_BUFFER_CACHE.with(|cache| {
+                cache.borrow_mut().reinit_cache(page_size);
+            });
+        }
+        if inner.page_arena.is_some() {
+            return Ok(pool.clone());
+        }
+        inner.db_page_size.store(page_size, Ordering::Relaxed);
+        inner.init_arenas()?;
+        Ok(pool.clone())
+    }
+
+    #[inline]
+    /// Marks the underlying pages for an `ArenaBuffer` on `Drop` as free and
+    /// available for use by another allocation.
+    pub fn free(&self, size: usize, arena_id: u32, page_id: u32) {
+        self.inner_mut().free(size, arena_id, page_id);
     }
 }
 
-#[cfg(test)]
-mod tests {
-    use super::*;
+impl PoolInner {
+    /// Allocate a buffer of the given length from the pool, falling back to
+    /// temporary thread local buffers if the pool is not initialized or full.
+    pub fn allocate(&self, len: usize) -> Buffer {
+        turso_assert!(len > 0, "Cannot allocate zero-length buffer");
 
-    fn is_send_sync_static<T: Send + Sync + 'static>() {}
+        let db_page_size = self.db_page_size.load(Ordering::Relaxed);
+        let wal_frame_size = db_page_size + WAL_FRAME_HEADER_SIZE;
 
-    #[test]
-    fn test_send_sync() {
-        is_send_sync_static::<BufferPool>();
+        // Check if this is exactly a WAL frame size allocation
+        if len == wal_frame_size {
+            let Some(wal_arena) = self.wal_frame_arena.as_ref() else {
+                return Buffer::new_temporary(len);
+            };
+            if let Some(FreeEntry { ptr, first_idx }) = wal_arena.try_alloc(len) {
+                tracing::trace!(
+                    "Allocated WAL frame buffer of length {} from arena {} at index {}",
+                    len,
+                    wal_arena.id,
+                    first_idx
+                );
+                return Buffer::new_pooled(ArenaBuffer::new(ptr, len, wal_arena.id, first_idx));
+            }
```

Comment:
Should this return a `new_temporary` at the end of this block instead of falling through to trying to allocate from page arena?